src/helpers/spotify_manager.ts

```diff
@@ -0,0 +1,237 @@
+import State from "../state";
+import _ from "lodash";
+import dbContext from "../database_context";
+import type QueriedSong from "../interfaces/queried_song";
+
+const logger = new IPCLogger("spotify_manager");
+
+const BASE_URL = "https://api.spotify.com/v1";
+
+interface SpotifyTrack {
+    name: string;
+    artists: Array<string>;
+}
+
+export interface PlaylistMetadata {
+    playlistID: string;
+    playlistName: string;
+    playlistLength: number;
+    matchedSongsLength: number;
+    thumbnailUrl?: string;
+}
+
+export interface MatchedPlaylist {
+    matchedSongs: Array<QueriedSong>;
+    metadata: PlaylistMetadata;
+}
+
+export default class SpotifyManager {
+    private accessToken: string;
+
+    constructor() {
+        this.refreshToken();
+    }
+
+    start(): void {
+        setInterval(async () => {
+            await this.refreshToken();
+        }, 3600000);
+    }
+
+    /**
+     * @param playlistID - The playlist to retrieve songs from
+     * @param isPremium - Whether the user is premium or not
+     *
+     */
+    getMatchedSpotifySongs = async (
+        playlistID: string,
+        isPremium: boolean
+    ): Promise<MatchedPlaylist> => {
+        if (
+            !process.env.SPOTIFY_CLIENT_ID ||
+            !process.env.SPOTIFY_CLIENT_SECRET
+        ) {
+            return {
+                metadata: {
+                    playlistID,
+                    playlistName: "",
+                    playlistLength: 0,
+                    matchedSongsLength: 0,
+                },
+                matchedSongs: [],
+            };
+        }
+
+        const spotifySongs: Array<SpotifyTrack> = [];
+        let requestURL = `${BASE_URL}/playlists/${playlistID}/tracks?${encodeURI(
+            "market=US&fields=items(track(name,artists(name))),next&limit=50"
+        )}`;
+
+        do {
+            try {
+                const response = (
+                    await Axios.get(requestURL, {
+                        headers: {
+                            Authorization: `Bearer ${this.accessToken}`,
+                            "Content-Type": "application/x-www-form-urlencoded",
+                        },
+                    })
+                ).data;
+
+                for (const item of response.items) {
+                    spotifySongs.push({
+                        name: item.track.name,
+                        artists: item.track.artists.map(
+                            (x: { name: string }) => x.name
+                        ),
+                    });
+                }
+
+                requestURL = response.next;
+            } catch (err) {
+                logger.error(`Failed fetching Spotify playlist. err = ${err}`);
+
+                if (err.response) {
+                    logger.info(err.response.data);
+                    logger.info(err.response.status);
+                }
+
+                break;
+            }
+        } while (requestURL);
```

Comment:
What are spotify rate limits like?

Reply:
They aren't well defined, but the response will contain a 429 and a `Retry-After` header. I'll check for that scenario and use our `retryJob` function